src/apify/scrapy/extensions/_httpcache.py

```diff
@@ -0,0 +1,190 @@
+from __future__ import annotations
+
+import gzip
+import io
+import pickle
+import re
+import struct
+from logging import getLogger
+from time import time
+from typing import TYPE_CHECKING
+
+from scrapy.http.headers import Headers
+from scrapy.responsetypes import responsetypes
+
+from apify import Configuration
+from apify.apify_storage_client import ApifyStorageClient
+from apify.scrapy._async_thread import AsyncThread
+from apify.storages import KeyValueStore
+
+if TYPE_CHECKING:
+    from scrapy import Request, Spider
+    from scrapy.http.response import Response
+    from scrapy.settings import BaseSettings
+    from scrapy.utils.request import RequestFingerprinterProtocol
+
+logger = getLogger(__name__)
+
+
+class ApifyCacheStorage:
+    """A Scrapy cache storage that uses the Apify `KeyValueStore` to store responses.
+
+    It can be set as a storage for Scrapy's built-in `HttpCacheMiddleware`, which caches
+    responses to requests. See HTTPCache middleware settings (prefixed with `HTTPCACHE_`)
+    in the Scrapy documentation for more information. Requires the asyncio Twisted reactor
+    to be installed.
+    """
+
+    def __init__(self, settings: BaseSettings) -> None:
+        self._expiration_max_items = 100
+        self._expiration_secs: int = settings.getint('HTTPCACHE_EXPIRATION_SECS')
+        self._spider: Spider | None = None
+        self._kvs: KeyValueStore | None = None
+        self._fingerprinter: RequestFingerprinterProtocol | None = None
+        self._async_thread: AsyncThread | None = None
+
+    def open_spider(self, spider: Spider) -> None:
+        """Open the cache storage for a spider."""
+        logger.debug('Using Apify key value cache storage', extra={'spider': spider})
+        self._spider = spider
+        self._fingerprinter = spider.crawler.request_fingerprinter
+        kvs_name = get_kvs_name(spider.name)
+
+        async def open_kvs() -> KeyValueStore:
+            config = Configuration.get_global_configuration()
+            if config.is_at_home:
+                storage_client = ApifyStorageClient.from_config(config)
+                return await KeyValueStore.open(name=kvs_name, storage_client=storage_client)
```

Comment:
This means indefinite data retention of this storage, which is undesirable in the case of a cache.

https://docs.apify.com/platform/storage/usage#data-retention

Reply:
Correct, that's why there's a "garbage collector" in `close_spider()`, which gradually removes expired items.

Reply:
That page mentions another interesting limit though:

> Storages' names can be up to 63 characters long.

Reply:
Okay, thanks.

> That page mentions another interesting limit though:
> Storages' names can be up to 63 characters long.

Maybe could you make sure in the `get_kvs_name` that the name isn't longer? I guess it's okay just slice off the rest.

Reply:
Done in [22af7ac](https://github.com/apify/apify-sdk-python/pull/403/commits/22af7acf189712d4e2ff7fb80a52c9af22f84d51)